### PR TITLE
Changed the way the rotation of a curve at a point is evaluated to match PathFollow2D

### LINKED ISSUE #78361: sample_baked_with_rotation gives wrong transform
### Godot version

4.0.3.stable

### System information

Windows 10

### Issue description

evaluating a curve's `sample_baked_with_rotation` gives wrong Transform2D:

https://github.com/godotengine/godot/assets/1028756/078552a4-3965-4d36-9446-022054ade672

The texture should read A B C D correctly, you can see it flips the texture vertically

It is trying to match textures *Positive Y* to the path direction, and then *positive x* to the right of the path.

I think the two correct options would be:
* map *negative y* (which would be *up*) to the path direction
* map *positive x* to the path direction (which I would like better, but would like to know your opinion)

I found this while working on this PR https://github.com/godotengine/godot/pull/77819
I could give a shot at fixing this during the weekend

### Steps to reproduce

Just run the example project

### Minimal reproduction project

[PathSampleBug.zip](https://github.com/godotengine/godot/files/11778276/PathSampleBug.zip)

### PR DESCRIPTION
**Pull request by botero-dev**
Fixes https://github.com/godotengine/godot/issues/78361

### CODE REVIEW

**Comment by AThousandShips** [comment]
This breaks `PathFollow2D` though, you need to update it as well, and the unit tests

**Comment by 0xafbf** [comment]
merging this would also benefit this other PR I'm working on https://github.com/godotengine/godot/pull/77819

**Comment by 0xafbf** [comment]
> This breaks `PathFollow2D` though, you need to update it as well, and the unit tests

You're right, I'll do that now

**Comment by 0xafbf** [comment]
I fixed the PathFollow2D code, but I have to leave now, I'll come back later to update tests and docs.

File: scene/2d/path_2d.cpp [hunk 1]
```diff
@@ -232,9 +232,8 @@ void PathFollow2D::_update_transform() {
 
 	if (rotates) {
 		Transform2D xform = c->sample_baked_with_rotation(progress, cubic);
-		xform.translate_local(v_offset, h_offset);
-		set_rotation(xform[1].angle());
-		set_position(xform[2]);
+		xform.translate_local(h_offset, v_offset);
+		set_transform(xform);
```
Comment (AThousandShips) [inline scene/2d/path_2d.cpp:236]: This makes it impossible to scale `PathFollow2D` which was previously possible

Comment (0xafbf) [reply scene/2d/path_2d.cpp]: Thanks for the insights! I was in a rush at that moment, but I think my fix now is cleaner.

**Comment by capnm** [comment]
Well, you need to tell your kids to paint rockets horizontally. In any case, better than backwards :)
LGTM

File: doc/classes/Curve2D.xml [hunk 2]
```diff
@@ -109,9 +109,12 @@
...
 				[codeblock]
-				var transform = curve.sample_baked_with_rotation(offset)
-				position = transform.get_origin()
-				rotation = transform.get_rotation()
+				var baked := curve.sample_baked_with_rotation(offset)
+				# transform can be set directly
+				transform = baked
+				# you can also read parts separately
+				position = baked.get_origin()
+				rotation = baked.get_rotation()
```
Comment (Calinou) [inline doc/classes/Curve2D.xml:117]:
```suggestion
				var baked := curve.sample_baked_with_rotation(offset)
				# The Transform2D can be set directly.
				transform = baked
				# You can also read the origin and rotation separately from the returned Transform2D.
				position = baked.get_origin()
				rotation = baked.get_rotation()
```

**Comment by Calinou** [comment]
I tested this PR locally (rebased on top of `master` f2ce0b6741049b69efb455394b8a161f68a468a6) with the MRP. While it fixes the issue mentioned in https://github.com/godotengine/godot/issues/78361, It breaks rendering of "fish bones" in the editor:

## This PR

image

## `master`

image

**Comment by capnm** [comment]
IIRC if I tested the initial commit the fish direction pfad was  was ok.

**Comment by capnm** [comment]
Screenshot from 2023-06-22 19-13-45

File: doc/classes/Curve2D.xml [hunk 3]
```diff
@@ -109,14 +109,12 @@
 			<description>
 				Similar to [method sample_baked], but returns [Transform2D] that includes a rotation along the curve, with [member Transform2D.origin] as the point position, [member Transform2D.x] as the sideways vector, and [member Transform2D.y] as the forward vector. Returns an empty transform if the length of the curve is [code]0[/code].
```
Comment (AThousandShips) [inline doc/classes/Curve2D.xml:110]: This is no longer the case with this PR and needs to be updated

**Review by kleonc — APPROVED** [review]
LGTM! :+1:

>Changed the way the rotation of a curve at a point is evaluated to match PathFollow2D

To precise, this PR rather makes `Curve2D::sample_baked_with_rotation`[^1] return a Transform2D compliant with the Godot's 2D coordinate system (previously it was returning a transform with the basis axes swapped). This implicates it makes it compliant with all other things compliant with the Godot's 2D coordinate system, including e.g. PathFollow2D.

[^1]: The changed `Curve2D::_sample_posture` is used internally only by `Curve2D::sample_baked_with_rotation`.

All usages of `Curve2D::sample_baked_with_rotation` within the codebase were changed appropriately (including more tests added).

The previous behavior was documented in #78362, so it can be considered a compatibility breaking change. I do consider this as a bugfix though (a buggy documented behavior is still a bug).
Sure there are probably some users using `Curve2D::sample_baked_with_rotation` and thus using some workarounds for the previous behavior. Their code would indeed break after merging this PR and thus they would need to get rid of the mentioned workarounds (shouldn't be a big deal).

So in case of merging this PR the change in `Curve2D::sample_baked_with_rotation`'s behavior needs to be clearly stated out in the changelog (or some other relevant place).

---

@0xafbf Rebasing would be probably a good idea.

File: doc/classes/Curve2D.xml [hunk 4]
```diff
@@ -107,16 +107,14 @@
 			<param index="0" name="offset" type="float" default="0.0" />
 			<param index="1" name="cubic" type="bool" default="false" />
 			<description>
-				Similar to [method sample_baked], but returns [Transform2D] that includes a rotation along the curve, with [member Transform2D.origin] as the point position, [member Transform2D.x] as the sideways vector, and [member Transform2D.y] as the forward vector. Returns an empty transform if the length of the curve is [code]0[/code].
+				Similar to [method sample_baked], but returns [Transform2D] that includes a rotation along the curve, with [member Transform2D.origin] as the point position and the [member Transform2D.x] vector pointing in the direction of the path at that point. Returns an empty transform if the length of the curve is [code]0[/code].
 				[codeblock]
-				var baked = curve.sample_baked_with_rotation(offset)
-				# This will rotate and position the node with the up direction pointing along the curve.
+				var baked := curve.sample_baked_with_rotation(offset)
```
Comment (akien-mga) [inline doc/classes/Curve2D.xml:112]: We don't typically use type hints in the documentation examples, to keep them simple.

**Comment by akien-mga** [comment]
Since this changes behavior and might surprise users, I'd prefer waiting for 4.3 so we have enough dev snapshots to get this tested and users to report any unexpected breakage in their projects.

**Comment by 0xafbf** [comment]
Rebased :) also took the colon from the docs example as requested.

Yeah, too bad this was approved a bit late for this release, but its great that its approved nonetheless :100:

Thanks for your efforts

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**